test/clj/rems/api/test_applications.clj

```diff
@@ -1962,6 +1962,73 @@
+            cat2 (test-helpers/create-catalogue-item! {:workflow-id wfid :resource-id res2})
+            app-id (test-helpers/create-application! {:actor applicant-id :catalogue-item-ids [cat1 cat2]})]
+        (testing "save partially filled duo codes"
+          (is (= {:success true}
+                 (send-command applicant-id
+                               {:type :application.command/save-draft
+                                :application-id app-id
+                                :field-values []
+                                :duo-codes [{:id "DUO:0000027" :restrictions [{:type :project :values [{:value "project id"}]}]}
+                                            {:id "DUO:0000007" :restrictions [{:type :mondo :values [{:id "MONDO:0015168"}]}]}]})))
+          (is (= {:duo/codes [{:id "DUO:0000027" :restrictions [{:type "project" :values [{:value "project id"}]}]}
+                              {:id "DUO:0000007" :restrictions [{:type "mondo" :values [{:id "MONDO:0015168"}]}]}]
+                  :duo/matches [{:id "DUO:0000024" :resource/id res1 :duo/valid? false}
+                                {:id "DUO:0000027" :resource/id res2 :duo/valid? "duo/needs-validation"}
+                                {:id "DUO:0000007" :resource/id res2 :duo/valid? true}]
+                  :duo/valid? false}
+                 (-> (get-application-for-user app-id applicant-id)
+                     :application/duo))))
+        (testing "save all required duo codes"
+          (is (= {:success true}
+                 (send-command applicant-id
+                               {:type :application.command/save-draft
+                                :application-id app-id
+                                :field-values []
+                                :duo-codes [{:id "DUO:0000024" :restrictions [{:type :date :values [{:value "2022-02-16"}]}]}
+                                            {:id "DUO:0000027" :restrictions [{:type :project :values [{:value "project id"}]}]}
+                                            {:id "DUO:0000007" :restrictions [{:type :mondo :values [{:id "MONDO:0015168"}]}]}]})))
+          (is (= {:duo/codes [{:id "DUO:0000024" :restrictions [{:type "date" :values [{:value "2022-02-16"}]}]}
+                              {:id "DUO:0000027" :restrictions [{:type "project" :values [{:value "project id"}]}]}
+                              {:id "DUO:0000007" :restrictions [{:type "mondo" :values [{:id "MONDO:0015168"}]}]}]
+                  :duo/matches [{:id "DUO:0000024" :resource/id res1 :duo/valid? true}
+                                {:id "DUO:0000027" :resource/id res2 :duo/valid? "duo/needs-validation"}
+                                {:id "DUO:0000007" :resource/id res2 :duo/valid? true}]
+                  :duo/valid? "duo/needs-validation"}
+                 (-> (get-application-for-user app-id applicant-id)
+                     :application/duo))))))
+    (testing "applicant fills fully valid duo codes"
```

Comment:
What happens if the applicant does not fill / send to one restriction but for the two others only?